src/main/java/org/opensearch/security/identity/SecurityTokenManager.java

```diff
@@ -0,0 +1,141 @@
+
+/**
+ * This class is the Security Plugin's implementation of the TokenManager used by all Identity Plugins.
+ * It handles the issuance of both Service Account Tokens and On Behalf Of tokens.
+ */
+public class SecurityTokenManager implements TokenManager {
+    private static final Logger logger = LogManager.getLogger(SecurityTokenManager.class);
+
+    private final ClusterService cs;
+    private final ThreadPool threadPool;
+    private final UserService userService;
+
+    private JwtVendor jwtVendor = null;
+    private ConfigModel configModel = null;
+
+    public SecurityTokenManager(final ClusterService cs, final ThreadPool threadPool, final UserService userService) {
+        this.cs = cs;
+        this.threadPool = threadPool;
+        this.userService = userService;
+    }
+
+    @Subscribe
+    public void onConfigModelChanged(final ConfigModel configModel) {
+        this.configModel = configModel;
+    }
+
+    @Subscribe
+    public void onDynamicConfigModelChanged(final DynamicConfigModel dcm) {
+        final Settings oboSettings = dcm.getDynamicOnBehalfOfSettings();
+        final Boolean enabled = oboSettings.getAsBoolean("enabled", false);
+        if (enabled) {
+            jwtVendor = createJwtVendor(oboSettings);
+        } else {
+            jwtVendor = null;
+        }
+    }
+
+    /** For testing */
+    JwtVendor createJwtVendor(final Settings settings) {
+        try {
+            return new JwtVendor(settings, Optional.empty());
+        } catch (final Exception ex) {
+            logger.error("Unable to create the JwtVendor instance", ex);
+            return null;
+        }
+    }
+
+    boolean oboNotSupported() {
+        return jwtVendor == null || configModel == null;
+    }
+
+    @Override
+    public ExpiringBearerAuthToken issueOnBehalfOfToken(final Subject subject, final OnBehalfOfClaims claims) {
+        if (oboNotSupported()) {
+            // TODO: link that doc!
+            throw new OpenSearchSecurityException(
+                "The OnBehalfOf token generation is not enabled, see {link to doc} for more information on this feature."
+            );
+        }
+
+        if (subject != null && !(subject instanceof NoopSubject)) {
+            logger.warn("Unsupported subject for OnBehalfOfToken token generation, {}", subject);
+            throw new IllegalArgumentException("Unsupported subject to generate OnBehalfOfToken");
+        }
+
+        if (Strings.isNullOrEmpty(claims.getAudience())) {
+            throw new IllegalArgumentException("Claims must be supplied with an audience value");
+        }
+
+        final User user = threadPool.getThreadContext().getTransient(ConfigConstants.OPENDISTRO_SECURITY_USER);
+        if (user == null) {
+            throw new OpenSearchSecurityException("Unsupported user to generate OnBehalfOfToken");
+        }
+
+        final TransportAddress callerAddress = null; /* OBO tokens must not roles based on location from network address */
```

Comment:
nit: the comment seems to be incomplete, probly a relic from previous changes? apologies if I didn't spot it earlier

Reply:
This was intentional - the address should be null

Reply:
Hmm, this is a pretty minor issue, I think it will be ok to leave it for now - whereas I've got fully passing CI